bundles/automation/org.eclipse.smarthome.automation.sample.handler.factories/src/main/java/org/eclipse/smarthome/automation/sample/handler/factories/SampleHandlerFactory.java

```diff
@@ -0,0 +1,163 @@
+import org.eclipse.smarthome.automation.type.ActionType;
+import org.eclipse.smarthome.automation.type.ConditionType;
+import org.eclipse.smarthome.automation.type.ModuleType;
+import org.eclipse.smarthome.automation.type.ModuleTypeRegistry;
+import org.eclipse.smarthome.automation.type.TriggerType;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.util.tracker.ServiceTracker;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Module Handler Factory sample implementation
+ *
+ * @author Vasil Ilchev - Initial Contribution
+ */
+public class SampleHandlerFactory implements ModuleHandlerFactory, ServiceTrackerCustomizer {
+    public static final String SUPPORTED_TRIGGER = "SampleTrigger";
+    public static final String SUPPORTED_CONDITION = "SampleCondition";
+    public static final String SUPPORTED_ACTION = "SampleAction";
+    public static final String MODULE_SPI_FACTORY_NAME = "[SampleHandlerFactory]";
+    public static final String UID_SEPARATOR = ":";
+    private BundleContext bc;
+    private Logger log;
+    private ModuleTypeRegistry moduleTypeRegistry;
+    private ServiceTracker moduleTypeRegistryTracker;
+    private ServiceReference moduleTypeRegistryRef;
+    private List<SampleTriggerHandler> createdTriggerHandler;
+    private static final Collection<String> types;
+    private ServiceRegistration serviceReg;
+
+    static {
+        List temp = new ArrayList();
+        temp.add(SUPPORTED_TRIGGER);
+        temp.add(SUPPORTED_CONDITION);
+        temp.add(SUPPORTED_ACTION);
+        types = Collections.unmodifiableCollection(temp);
+    }
+
+    public SampleHandlerFactory(BundleContext bc) {
+        this.bc = bc;
+        moduleTypeRegistryTracker = new ServiceTracker(bc, ModuleTypeRegistry.class.getName(), this);
+        moduleTypeRegistryTracker.open();
+        log = LoggerFactory.getLogger(SampleHandlerFactory.class);
+        createdTriggerHandler = new ArrayList<SampleTriggerHandler>();
+        serviceReg = bc.registerService(ModuleHandlerFactory.class.getName(), this, null);
+    }
+
+    protected void disposeHandler(ModuleHandler handler) {
+        createdTriggerHandler.remove(handler);
+    }
+
+    @Override
+    public Object addingService(ServiceReference reference) {
+        Object result = null;
+        if (moduleTypeRegistryRef == null && reference != null) {
+            result = moduleTypeRegistry = (ModuleTypeRegistry) bc.getService(reference);
+        }
+        return result;
+    }
+
+    @Override
+    public void modifiedService(ServiceReference reference, Object service) {
+        // do nothing
+    }
+
+    @Override
+    public void removedService(ServiceReference reference, Object service) {
+        if (moduleTypeRegistry == service) {
+            // dispose();
```

Comment:
It will dispose resources related to the moduleTypeAdmin service. It must not close the ModuelTypeTracker. The triacker has to be closed only on bundle stop. It is fixed.